docs/contributors/code/e2e/README.md

```diff
@@ -0,0 +1,131 @@
+
+```bash
+# Run all available tests.
+xvfb-run npm run test:e2e:playwright
+
+# Only run webkit tests.
+xvfb-run -- npm run test:e2e:playwright -- --project=webkit
+```
+
+## Best practices
+
+<details>
+	<summary><h3>Forbid `$`, use `locator` instead</h3></summary>
+
+In fact, any API that returns `ElementHandle` is [discouraged](https://playwright.dev/docs/api/class-page#page-query-selector). This includes `$`, `$$`, `$eval`, `$$eval`, etc. [`Locator`](https://playwright.dev/docs/api/class-locator) is a much better API and can be used with playwright's [assertions](https://playwright.dev/docs/api/class-locatorassertions). This also works great with Page Object Model since that locator is lazy and doesn't return a promise.
+</details>
+
+<details>
+	<summary><h3>Use accessible selectors</h3></summary>
+
+Use the selector engine [role-selector](https://playwright.dev/docs/selectors#role-selector) to construct the query wherever possible. It enables us to write accessible queries without having to rely on internal implementations. The syntax should be straightforward and looks like this:
+
+```js
+// Select a button with the accessible name "Hello World" (case-insensitive).
+page.locator( 'role=button[name="Hello World"i]' );
+
+// Using short-form API, the `name` is case-insensitive by default.
+page.getByRole( 'button', { name: 'Hello World' } );
+```
+
+It's recommended to append `i` to the name attribute to match it case-insensitively wherever it makes sense. It can also be chained with built-in selector engines to perform complex queries:
+
+```js
+// Select a button with a name ends with `Back` and is visible on the screen.
+page.locator( 'role=button[name=/Back$/] >> visible=true' );
+// Select a button with the (exact) name "View options" under `#some-section`.
+page.locator( 'css=#some-section >> role=button[name="View options"]' );
+```
+
+See the [official documentation](https://playwright.dev/docs/selectors#role-selector) for more info on how to use them.
+</details>
+
+<details>
+	<summary><h3>Selectors are strict by default</h3></summary>
+
+To encourage better practices for querying elements, selectors are [strict](https://playwright.dev/docs/api/class-browser#browser-new-page-option-strict-selectors) by default, meaning that it will throw an error if the query returns more than one element.
+</details>
+
+<details>
+	<summary><h3>Don't overload test-utils, inline simple utils</h3></summary>
+
+`e2e-test-utils` are too bloated with too many utils. Most of them are simple enough to be inlined directly in tests. With the help of accessible selectors, simple utils are easier to write now. For utils that only take place on a certain page, use Page Object Model instead (with an exception of clearing states with `requestUtils` which are better placed in `e2e-test-utils`). Otherwise, only create an util if the action is complex and repetitive enough.
```

Comment:
```suggestion
`e2e-test-utils` are too bloated with too many utils. Most of them are simple enough to be inlined directly in tests. With the help of accessible selectors, simple utils are easier to write now. For utils that only take place on a certain page, use Page Object Model instead (with an exception of clearing states with `requestUtils` which are better placed in `e2e-test-utils`). Otherwise, only create a util if the action is complex and repetitive enough.
```